File: C++/3.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        """
        :type: s: str
        :rtype: int
        """
       int left = 0, right = 0;
       int max_length = 0;
       set<int> used_char;
       set<int>::iterator iter;
       while (right < s.length()) {
           iter = used_char.find(s[right]);
           if (iter == used_char.end()) {
               used_char.insert(s[right]);
               right += 1;
               max_length = max(max_length, right - left);
           }
           else {
               used_char.erase(s[left]);
               left += 1;
           }
       }
       return max_length;
    }
};

File: Python/3.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        left, right = 0, 0 # sliding window
        max_length = 0
        used_char = set() # like dictionary, but without value
        while right < len(s):
            if s[right] not in used_char:
                used_char.add(s[right])
                right += 1
                max_length = max(max_length, right-left) # when right += 1, the max_length may update, but when left += 1, the max_length won't update. 
            else:
                used_char.remove(s[left]) # keep remove char until there is no duplicate char in the sliding window 
                left += 1
        return max_length